addons/binding/org.openhab.binding.smappee/src/main/java/org/openhab/binding/smappee/handler/SmappeeHandler.java

```diff
@@ -0,0 +1,185 @@
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "Check configuration, Client secret must be provided");
+            return;
+        }
+        if (config.username.isEmpty()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "Check configuration, Username must be provided");
+            return;
+        }
+        if (config.password.isEmpty()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "Check configuration, Password must be provided");
+            return;
+        }
+        if (config.serviceLocationName.isEmpty()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "Check configuration, Service location name must be provided");
+            return;
+        }
+
+        logger.debug("Initialize Network handler.");
+
+        smappeeService = new SmappeeService(config);
+
+        // contact Smappee API
+        scheduler.submit(new Runnable() {
+            @Override
+            public void run() {
+                initializeSmappeeService();
+            }
+        });
+    }
+
+    private void initializeSmappeeService() {
+        try {
+            smappeeService.initialize();
+        } catch (CommunicationException ex) {
+
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "Could not contact Smappee, retrying");
+
+            // try again in 30 seconds
+            scheduler.schedule(new Runnable() {
+                @Override
+                public void run() {
+                    initializeSmappeeService();
+                }
+            }, 30, TimeUnit.SECONDS);
+
+            return;
+        } catch (InvalidConfigurationException ex) {
+            // bad configuration, stay offline until user corrects the configuration
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, ex.getMessage());
+
+            return;
+        }
+
+        // ok, initialization succeeded
+        smappeeService.startAutomaticRefresh(scheduler, this);
+
+        updateStatus(ThingStatus.ONLINE);
+
+        logger.debug("Initialize discovery service.");
+        registerDeviceDiscoveryService();
+        discoveryService.startScan(null);
+    }
+
+    @Override
+    public void dispose() {
+        if (smappeeService != null) {
+            smappeeService.stopAutomaticRefresh();
+        }
+
+        if (discoveryService != null) {
+            discoveryService.stopScan();
+            unregisterDeviceDiscoveryService();
+        }
+    }
+
+    private void registerDeviceDiscoveryService() {
+        discoveryService = new SmappeeDiscoveryService(this.smappeeService, this.thing.getUID());
+        discoveryServiceRegistration = bundleContext.registerService(DiscoveryService.class.getName(), discoveryService,

```

Comment:
You must not access the `bundleContext` from a handler.
The recommended way of having a discovery service that works on a bridge is to instantiate it in the handler factory and pass the handler to the discovery service instance as a constructor parameter. See the [hue binding](https://github.com/eclipse/smarthome/blob/master/extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/HueThingHandlerFactory.java#L92-L109) as an example.

Reply:
@kaikreuzer 

Philips hue does the same thing ?

private synchronized void registerLightDiscoveryService(HueBridgeHandler bridgeHandler) {
        HueLightDiscoveryService discoveryService = new HueLightDiscoveryService(bridgeHandler);
        discoveryService.activate();
        this.discoveryServiceRegs.put(bridgeHandler.getThing().getUID(), **bundleContext
                .registerService(DiscoveryService.class.getName(), discoveryService, new Hashtable<String, Object>()));**
    }

Reply:
It's the use of `bundleContext` (and related code)  should not be done in the `SmappeeHandler`, but in the Factory class. The hue implementation is in the `HueThingHandlerFactory`.

Reply:
ok, I will refactor this

Reply:
Done